pkg/db/db.go

```diff
@@ -0,0 +1,319 @@
+	}
+
+	if db == nil {
+		panic("nil db")
+	}
+
+	err := db.Update(func(tx *bolt.Tx) error {
+		for _, root := range roots {
+			_, err := tx.CreateBucketIfNotExists([]byte(root))
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &DB{db: db}, nil
+}
+
+func (d *DB) Close() {
+	d.db.Close()
+}
+
+const marshalledKey = "marshalled"
+
+type Model interface {
+	DBIndexes() (map[string]string, map[string]bool, map[string]int)
+	DBRoot() string
+	DBUnmarshal(data []byte) (Model, error)
+	DBMarshal() ([]byte, error)
+	DBKey() string
+}
+
+func (d *DB) Save(m Model) error {
+	key := m.DBKey()
+	if key == "" {
+		return fmt.Errorf("save failed, model key nil")
+	}
+
+	strings, bools, ints := m.DBIndexes()
+	var err error
+
+	for k, v := range bools {
+		if strings[k] != "" {
+			return fmt.Errorf("model misconfigured, index keys overlap across type")
+		}
+		strings[k] = strconv.FormatBool(v)
+	}
+
+	for k, v := range ints {
+		if strings[k] != "" {
+			return fmt.Errorf("model misconfigured, index keys overlap across type")
+		}
+		strings[k] = strconv.Itoa(v)
+	}
+
+	marshalled, err := m.DBMarshal()
+	if err != nil {
+		return err
+	}
+
+	err = d.db.Update(func(tx *bolt.Tx) error {
+		rootb := tx.Bucket([]byte(m.DBRoot()))
+		mb, err := rootb.CreateBucketIfNotExists([]byte(m.DBKey()))
+		if err != nil {
+			return err
+		}
+		err = mb.Put([]byte(marshalledKey), marshalled)
+		if err != nil {
+			return err
+		}
+		for k, v := range strings {
+			err = mb.Put([]byte(k), []byte(v))
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func Q() Query {
+	return Query{}
+}
+
+type Query struct {
+	Strings map[string]string
+	Bools   map[string]bool
+	Ints    map[string]int
+}
+
+func (q Query) String(k string, v string) Query {
+	q.Strings[k] = v
+	return q
+}
+
+func (q Query) Bool(k string, v bool) Query {
+	q.Bools[k] = v
+	return q
+}
+
+func (q Query) Int(k string, v int) Query {
+	q.Ints[k] = v
+	return q
+}
+
+type queryResult struct {
+	key        string
+	marshalled []byte
+}
+
+func (d *DB) Get(m Model, query Query) ([]Model, error) {
+	normalizedQueryValues, err := validateAndNormalize(m, query)
+	if err != nil {
+		return nil, err
+	}
+
+	var marshalled []queryResult
+
+	err = d.db.View(func(tx *bolt.Tx) error {
+		rootb := tx.Bucket([]byte(m.DBRoot()))
+		_ = rootb.ForEach(func(k, v []byte) error {
+			rootbentryb := rootb.Bucket(k)
+			if rootbentryb == nil {
+				return nil
+			}
+			var match = true
+			for queryKey, queryVal := range normalizedQueryValues {
+				storedVal := rootbentryb.Get([]byte(queryKey))
+				if string(storedVal) != queryVal {
+					match = false
+					break
+				}
+			}
+			if match {
+				marshalled = append(marshalled, queryResult{key: string(k), marshalled: rootbentryb.Get([]byte(marshalledKey))})
```

Comment:
```suggestion
				marshalled = append(marshalled, rootbentryb.Get([]byte(marshalledKey))...)
				marshalled = append(marshalled, []byte(",")...)
				
```

and then sometime later:

```
marshalled = append(marshalled, []byte("]")...)
```

Reply:
oh i dont like that at alllllll lol - took me forever to figure out what you were suggesting; i think clarity should win out

Reply:
@jkminneti you're right that its heinous. but it would make the outside appear more idiomatic. You also would only ever do a single unmarshal, not sure if there any performance considerations there (not that I think we should be considering performance at this stage)